### PR TITLE
Integration LightwaveRF switches

### PR DESCRIPTION
**Pull request by max246**
# What does this implement/fix?

This is an implemantion of [LightwaveRF](https://lightwaverf.com/) which allows to dump and send raw messages

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2915

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ x ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml

lightwaverf:
  read_pin: 13
  write_pin: 14


switch:
  - platform: template
    name: "Turn off sofa"
    id: light_off_ceiling_sofa
    turn_on_action:
      lightwaverf.send_raw:
        code:  [0x04, 0x00, 0x00, 0x00, 0x0f, 0x03, 0x0d, 0x09, 0x08, 0x08]
        name: "Sofa"
        repeat: 1
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by descipher** [comment]
Looks good to me.

**Comment by max246** [comment]
@jesserockz  could this be approved?

File: esphome/components/lightwaverf/__init__.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            raise cv.Invalid("One or more values in the code are negative")
+
+    return value
+
+
+
+
+
+
+"""
```
Comment (jesserockz) [inline esphome/components/lightwaverf/__init__.py]: Commented code is not allowed

**Review by jesserockz — CHANGES_REQUESTED** [review]
Please remove the EEPROM stuff, if you want to store state etc, then you can use `global_preferences->make_preference(...)` See `esphome/core/preferences.h`

**Comment by max246** [comment]
> Please remove the EEPROM stuff, if you want to store state etc, then you can use `global_preferences->make_preference(...)` See `esphome/core/preferences.h`

I am going to remove the eeprom stuff which was part of an old library and I was meant to remove it.

I have to re-test the validation code before uncomment it, I totally forgot about it :(

File: esphome/components/lightwaverf/__init__.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+def validate_rc_switch_raw_code(value):
+    if not isinstance(value, list):
+        raise cv.Invalid("All Lightwave rf raw codes must a list of hex (0x00,0x00)")
+
+    not_valid_value = any(i for i, val in enumerate(value) if val < 0)
+    if not_valid_value:
+        raise cv.Invalid("One or more values in the code are negative")
+
+    return value
```
Comment (jesserockz) [inline esphome/components/lightwaverf/__init__.py]: Im pretty sure the existing validation (`[cv.Any(cv.hex_uint8_t)]`) should prevent this custom validation form being necessary.

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/lightwaverf/LwRx.h [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  bool rx_translate = true;     // Set false to get raw data
+
+  uint8_t rx_state = 0;
+
+  uint8_t rx_num_bits = 0;   // number of bits in the current uint8_t
+  uint8_t rx_num_bytes = 0;  // number of bytes received
+
+  uint16_t lwrx_stats[RX_STAT_COUNT];  // unsigned int
+
+  bool lwrx_stats_enable = true;
```
Comment (kbx81) [inline esphome/components/lightwaverf/LwRx.h:118]: Are you sure you want _all_ of these to be `public`?

File: esphome/components/lightwaverf/LwTx.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+  // Gap multiplier byte is used to multiply gap if longer periods are needed for experimentation
+  // If gap is 255 (35msec) then this to give a max of 9 seconds
+  // Used with low repeat counts to find if device times out
+  uint8_t tx_gap_multiplier = 0;  // Gap extension byte
+
+  uint8_t tx_bit_mask = 0;   // bit mask in current byte
+  uint8_t tx_num_bytes = 0;  // number of bytes sent
+
+  InternalGPIOPin *tx_pin;
```
Comment (kbx81) [inline esphome/components/lightwaverf/LwTx.h:84]: As with the receiver, should these _all_ really be `public`?

File: esphome/components/lightwaverf/LwRx.cpp [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+  rx_timeout = timeout;
+}
+
+/**
+  Add a pair to filter received messages
+  pairdata is device,dummy,5*addr,room
+  pairdata is held in translated form to make comparisons quicker
+**/
+uint8_t LwRx::lwrx_addpair(const uint8_t *pairdata) {
+  if (rx_paircount < RX_MAXPAIRS) {
```
Comment (kbx81) [inline esphome/components/lightwaverf/LwRx.cpp]:
```suggestion
  if (this->rx_paircount < RX_MAXPAIRS) {
```

It seems you've got a bunch of places throughout where you're referring to a member variable without the `this->` prefix. We tend to prefer this is appended to each use of class variables within the code base. Please fix this for all member variable uses throughout. 😇

**Review by kbx81 — CHANGES_REQUESTED** [review]

**Comment by max246** [comment]
@kbx81 due to the interrupt system, these variables need to be outside. 
I think the PR is ready to merge

**Comment by max246** [comment]
@jesserockz any decision on this? :)

File: esphome/components/lightwaverf/LwRx.h (hunk 3 above)
Comment (jesserockz) [reply esphome/components/lightwaverf/LwRx.h:118]: While these are public here, this is a instance that is only protected in the main component. This should be fine.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**